enhancements/insights/pulling-pull-secret-from-ocm.md

```diff
@@ -0,0 +1,168 @@
+which involves initiating the transfer from the OpenShift Cluster Manager (OCM) API
+and then updating the global `pull-secret` on the cluster with the new owner’s `pull-secret`.
+If the `pull-secret` is not changed on the cluster within 5 days, the ownership transfer is canceled
+and needs to be initiated again.
+This process applies to all scenarios that require cluster ownership transferred across Red Hat accounts.
+
+The intent is to simplify the cluster ownership process to a one step process.
+
+## Motivation
+
+The entry point to many OpenShift subscriptions is for a user to start with a 60 day trial from the OCM with personal credentials
+and then migrate it over to a company or organization account. In addition, reports from the field indicate that customers in general when transferring ownership, either follow one of the steps in the process and do not follow through with the other step required to complete the transfer. This enhancement will make the transfer frictionless and provide a better user experience.
+
+
+### Goals
+The goal of this enhancement is to describe the API and the interaction between OCM and the cluster that is being transferred to update the `pull-secret` on that cluster:
+
+- Extend the Insights Operator config with a new OCM API endpoint to be able to query the new `pull-secret` content
+- Periodically (once a day) query the new endpoint and if there is a new or updated content then update the `pull-secret`
+
+### Non-Goals
+- The enhancement will not describe the OCM user experience.
+- The enhancement will not provide a generic way to pull information from OCM into the cluster.
+- The process will not apply to disconnected clusters.
+
+## Proposal
+
+The Insights operator already communicates with the OCM API (see [Insights Operator pulling and exposing SCA certs from the OCM API](../insights/pulling-sca-certs-from-ocm.md). This is the proposed flow for updating the `pull-secret`:
+
+1. OCM provides an API that the cluster can call to check the current global `pull-secret`.
+2. User starts the migration process in the OCM by providing the `pull-secret` associated with the account to which the ownership will be transferred to.
+   * The details of whether the user provides the pull secret as a means of verification, or whether a request/approval workflow is implemented, is left to the OCM/UI flow. What is pertinent here is that there is an endpoint available to a cluster entity to fetch the new `pull-secret`.
+3. OCM validates the `pull-secret` for the new account and makes it available to the cluster for some time period i.e. the new API will now return the new `pull-secret` provided by the user.
+4. The Insights operator calls this API periodically, fetches the `pull-secret` and compares it with the current global `pull-secret` on the cluster.
+5. If there is a mismatch, the Insights operator applies the new `pull-secret`.
```

Comment:
This makes me nervous.  The cluster's global pull secret is [explicitly doc'ed as admin-editable][1].  And removing pull-secret entries is [explicitly doc'ed as a way to disable remote monitoring][2].  This line will have the Insights operator and the local cluster-admin sharing responsibility for the global pull secret, and sharing is hard ;).  How will Insights distinguish entries that the local admin has appended?  Perhaps by assuming that RH-supplied pull secrets will never drop entries for a particular domain?  How will Insights distinguish entries that the local admin has explicitly removed?  Perhaps by using the OCM-API token alone as a marker for "there's a mismatch" and then only updating tokens for domains that exist in the current global pullspec?  Even if all that gets sorted, this is still granting the Insights operator huge power to unilaterally break the cluster (e.g. if someone manages to compromise or MitM the OCM-API and recommend a pivot to a garbage secret).  I think this needs a very detailed entry in the risks and mitigation section.

[1]: https://docs.openshift.com/container-platform/4.9/openshift_images/managing_images/using-image-pull-secrets.html#images-update-global-pull-secret_using-image-pull-secrets
[2]: https://docs.openshift.com/container-platform/4.9/support/remote_health_monitoring/opting-out-of-remote-health-reporting.html#insights-operator-new-pull-secret_opting-out-remote-health-reporting

Reply:
My understanding is that the API will serve only the updated `pull-secret` (the update is a user action in the OCM IMO) and only within the given window. This should minimize the potential diffs between the secrets content. But yes, I agree that this should be described carefully. 
The only mitigation (in similar disaster scenarios) I can think of is to refer to the documentation and manual steps how to try again or update the secret manually. Are there any better ideas?

Reply:
is there a clear boundary for "the global pull secret provided by the OCM"?  There's a single global pull secret in-cluster, and the bits we hand out as an OCP pull secret could be some, all, or none of the cluster's current pull secret.  For example, maybe the cluster admin added a docker.io pull secret to the cluster's global secret.  We wouldn't want to clobber that.

Reply:
@wking I added the following sentence there:

> Insights operator only replaces the entries provided in the updated `pull-secret` from the OCM API.

It means that the Insights operator cannot add or delete anything from the secret (I can try to emphasize this even more). It can only replace existing entries with new ones and if the entries (entry names) don't match then we should inform use (probably alert?)

Reply:
Probably also means that no operator besides Insights can use "remove an entry from the pull secret" as a knob for turning things off.

Reply:
Yes, is there such a workflow today?

The one I know of is disabling the remote health functionality https://docs.openshift.com/container-platform/4.9/support/remote_health_monitoring/opting-out-of-remote-health-reporting.html.

If the remote health functionality is disabled, this workflow would not work at all, so implementaiton-wise probably not a worry, we should however make sure the docs are updated to ensure people realize this functionality would be lost as well.

Are there other cases where people might edit the relevant parts of pull secret outside of the knowledge of OCM?

Reply:
We discussed this yesterday and I updated the enhancement.